task/pa11y.js

```diff
@@ -23,60 +23,49 @@ exports.runPa11yOnTasks = runPa11yOnTasks;
 
 // Initialise the task
 function initTask(config, app) {
-
 	if (!config.cron) {
 		config.cron = '0 30 0 * * *'; // 00:30 daily
 	}
-	const job = new CronJob(config.cron, run.bind(null, app));
+	const job = new CronJob(config.cron, taskRunner.bind(null, app));
 	job.start();
-
 }
 
-// Run the task
-function run(app) {
-
+// Runs the task
+async function taskRunner(app) {
 	console.log('');
-	console.log(chalk.grey('Starting to run tasks @ %s'), new Date());
-
-	async.waterfall([
-
-		next => app.model.task.getAll(next),
-
-		(tasks, next) => runPa11yOnTasks(tasks, app, next)
-
-	], error => {
-
-		if (error) {
-			console.error(chalk.red('Failed to run tasks: %s'), error.message);
-			console.log('');
-			process.exit(1);
-		}
-
-	});
-
+	console.log(chalk.grey('Starting to run task @ %s'), new Date());
+
+	try {
+		const tasks = await app.model.task.getAll();
+		runPa11yOnTasks(tasks, app);
```

Comment:
I think that as `runPa11yOnTasks` is using `async.queue` which in turn is using `setImmediate` internally, the try/catch won't capture the error.

`async/queue` has a `.error` handler though. `runPa11yOnTasks` would need to be changed to return a Promise which the `.error` handler would reject. Then `await runPa11yOnTasks` in this try/catch.

http://caolan.github.io/async/v3/docs.html#queue

Reply:
Given that this broke and we had no idea, we should really add some unit tests for this. It shouldn't be too bad as the `app` is passed in so the model can have `task.getAll` and `task.runById` mocked out.

We'd need a cron string to make it run every 1 sec or something. Probably wouldn't be the prettiest looking test....

Reply:
You're right in that this needs tests as none of this code is being tested at the moment, which is why we didn't realise it's broken.

That said, we're trying to get a fix released ASAP as dashboard is broken because of this so we probably want to add the tests later. Jon has been looking at them already.

I'm not familiar with promises, but I will have a look now and see if I can make it work that way.

Reply:
Joey you are correct, but 😁 
-  `runPa11yOnTasks` is exported by this module so we were not keen on changing what it returns, as we don't know what other code uses it as yet (without more refactoring).
- the try/catch should never catch _as long as runPa11yOnTasks doesn't change_. This is "illogical" or belt & braces, depending on ones opinion. It also reproduces the original behaviour of the code IIRC which we didn't want to change as this isn't adequately tested yet.

Reply:
Yeah, fair enough. I've created #94 to add osme unit tests around this so we have a bit of stability.

Reply:
Thanks everyone! Can we merge then?